plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,174 @@
+import org.jetbrains.kotlin.idea.refactoring.introduce.selectElementsWithTargetSibling
+import org.jetbrains.kotlin.idea.refactoring.introduce.showErrorHint
+import org.jetbrains.kotlin.idea.refactoring.introduce.showErrorHintByKey
+import org.jetbrains.kotlin.idea.refactoring.introduce.validateExpressionElements
+import org.jetbrains.kotlin.idea.util.psi.patternMatching.toRange
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.plainContent
+import org.jetbrains.kotlin.resolve.constants.evaluate.ConstantExpressionEvaluator
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+class KotlinIntroduceConstantHandler(
+    val helper: ExtractionEngineHelper = InteractiveExtractionHelper
+) : RefactoringActionHandler {
+    object InteractiveExtractionHelper : ExtractionEngineHelper(INTRODUCE_CONSTANT) {
+        private fun getExtractionTarget(descriptor: ExtractableCodeDescriptor) =
+            propertyTargets.firstOrNull { it.isAvailable(descriptor) }
+
+        override fun validate(descriptor: ExtractableCodeDescriptor) =
+            descriptor.validate(getExtractionTarget(descriptor) ?: ExtractionTarget.FUNCTION)
+
+        override fun configureAndRun(
+            project: Project,
+            editor: Editor,
+            descriptorWithConflicts: ExtractableCodeDescriptorWithConflicts,
+            onFinish: (ExtractionResult) -> Unit
+        ) {
+            val descriptor = descriptorWithConflicts.descriptor
+            val target = getExtractionTarget(descriptor)
+            if (target != null) {
+                val options = ExtractionGeneratorOptions(target = target, delayInitialOccurrenceReplacement = true, isConst = true)
+                doRefactor(ExtractionGeneratorConfiguration(descriptor, options), onFinish)
+            } else {
+                showErrorHint(
+                    project,
+                    editor,
+                    KotlinBundle.message("error.text.can.t.introduce.constant.for.this.expression"),
+                    INTRODUCE_CONSTANT
+                )
+            }
+        }
+    }
+
+    fun doInvoke(project: Project, editor: Editor, file: KtFile, elements: List<PsiElement>, target: PsiElement) {
+        val adjustedElements = (elements.singleOrNull() as? KtBlockExpression)?.statements ?: elements
+        when {
+            adjustedElements.isEmpty() -> {
+                showErrorHintByKey(
+                    project, editor, "cannot.refactor.no.expression",
+                    INTRODUCE_CONSTANT
+                )
+            }
+            else -> {
+                val options = ExtractionOptions(extractAsProperty = true)
+                val extractionData = ExtractionData(file, adjustedElements.toRange(), target, null, options)
+                ExtractionEngine(helper).run(editor, extractionData) {
```

Comment:
I tried around a bit with `project.executeCommand` with `runWriteAction` inside, but I did not manage to get the wanted behaviour. The code in `KotlinIntroduceVariableHandler´ looks completly different from `KotlinIntroducePropertyHandler`, so I don't know what exactly I have to do.

`KotlinIntroducePropertyHandler` (CTRL + ALT + F) has the same issue, since I oriented myself here